### ISSUE TITLE
Is 0x80000000 a valid error code?

### ISSUE THREAD

**Issue by gjasny** [openssl/openssl#18785]
Hi,

while porting our code from OpenSSL 1.1.1 to 3.0.0 I came across some interesting behavior in the error handling.

Testcase (Windows 10 21H1, Visual Studio 2022, OpenSSL 3.0.5):
```c++
#include <iostream>

#include <openssl/err.h>
#include <openssl/ssl.h>

// see https://github.com/chriskohlhoff/asio/blob/9902fe7d1d6b5321f4ab2f4ddbfd19cf39160891/asio/include/asio/ssl/impl/context.ipp#L541-L554
void load_verify_file(SSL_CTX* ctx, const char* file)
{
    ::ERR_clear_error();

    if (::SSL_CTX_load_verify_file(ctx, file) != 1) {

        const char *file, *func, *data;
        int line, flags;
        auto e = ERR_peek_error_all(&file, &line, &func, &data, &flags);

        std::cout << "e=0x" << std::hex << e << '\n';
        std::cout << " file=" << std::dec << file << '\n';
        std::cout << " line=" << std::dec << line << '\n';
        std::cout << " func=" << std::dec << func << '\n';
        std::cout << " data=" << std::dec << data << '\n';
        std::cout << " flags=" << std::dec << flags << '\n';
    }
}

int main() {

    const SSL_METHOD* m = TLS_method();

#if 0
    SSL_CTX* ctx1 = SSL_CTX_new(m);
    load_verify_file(ctx1, "D:/Git/rtc-platform/modules/network/Test/Data/test_trustedroots.pem");
    SSL_CTX_free(ctx1);
#endif

    SSL_CTX* ctx2 = SSL_CTX_new(m);
    load_verify_file(ctx2, "");
    SSL_CTX_free(ctx2);

    return 0;
}
```

If the `SSL_CTX_load_verify_file` function was successful before (even for a different context) its return value is `0` and the error code is `0x80000000`. If I remove initialization of ctx1 the error code changes to `0x80000003`.

With ctx1:
```
e=0x80000000
 file=crypto\bio\bss_file.c
 line=67
 func=BIO_new_file
 data=calling fopen(, r)
 flags=3
```

Without ctx1:
```
e=0x80000003
 file=crypto\bio\bss_file.c
 line=67
 func=BIO_new_file
 data=calling fopen(, r)
 flags=3
```

The problem is that ASIO [translates](https://github.com/chriskohlhoff/asio/blob/9902fe7d1d6b5321f4ab2f4ddbfd19cf39160891/asio/include/asio/ssl/impl/context.ipp#L1301-L1314) `0x80000000` error into an boost error:

```c++
asio::error_code context::translate_error(long error)
{
#if (OPENSSL_VERSION_NUMBER >= 0x30000000L)
  if (ERR_SYSTEM_ERROR(error))
  {
    return asio::error_code(
        static_cast<int>(ERR_GET_REASON(error)),
        asio::error::get_system_category());
  }
#endif // (OPENSSL_VERSION_NUMBER >= 0x30000000L)

  return asio::error_code(static_cast<int>(error),
      asio::error::get_ssl_category());
}
```

And `0x80000000` maps to the reason `0` which unfortunately means "No error" in the asio context.

Could you please add your thoughts about the issue?

1. Why does the error code issued by `SSL_CTX_load_verify_file` change
2. Is returning a reason of `0` a bug?

Thanks,
Gregor

**Comment by slontis** [comment]
If you look at the line of code that it points to it does
ERR_raise_data(ERR_LIB_SYS, get_last_sys_error(),

So its doing some sort of call to a fopen that fails and then it tries to use the system error result (whatever that is)..

**Comment by mattcaswell** [comment]
Right - as @slontis says this is raised by:

````
        ERR_raise_data(ERR_LIB_SYS, get_last_sys_error(),
                       "calling fopen(%s, %s)",
                       filename, mode);
````

So, it seems that `get_last_sys_error()` is returning 0. We really shouldn't be raising an error code with a 0 reason code. Perhaps we should translate a 0 reason code into an "internal error" reason code or something.

A separate question is why `get_last_sys_error()` is returning 0 when `fopen` has failed.

**Comment by t8m** [comment]
> We really shouldn't be raising an error code with a 0 reason code. Perhaps we should translate a 0 reason code into an "internal error" reason code or something.

Yeah, there should be an "unknown sys error reason code" or something like that.

**Comment by gjasny** [comment]
The reason why the error code is zero is the weird `GetLastError()` / `errno` handling on Windows. The fopen-family of functions only sets `errno` on error and not the "last" error (from [here](https://github.com/openssl/openssl/blob/1efd8533e1ccc5c5e69795eb393a6b79b62e48e2/crypto/o_fopen.c#L70)):

```
_wfopen: file=0000000000000000 errno=22 GetLastError()=0
```

`BIO_new_file`, the caller of `openssl_fopen` has no idea if it got a null pointer due to `MultiByteToWideChar` (which sets last error) or `fopen` which sets `errno`.

To catch all errors like this the "unknown sys error reason code" approach as suggested by @t8m seems to be most suited.
